controllers/idle_handlers.go

```diff
@@ -22,29 +22,114 @@ import (
 	"os"
 	"path/filepath"
 
+	"github.com/openshift-kni/lifecycle-agent/ibu-imager/ops"
+	"github.com/openshift-kni/lifecycle-agent/internal/backuprestore"
 	"github.com/openshift-kni/lifecycle-agent/internal/common"
 
 	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
 	"github.com/openshift-kni/lifecycle-agent/controllers/utils"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-func (r *ImageBasedUpgradeReconciler) cleanupUnbootedStateroot(stateroot string) error {
+func (r *ImageBasedUpgradeReconciler) handleAbort(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	r.Log.Info("Starting handleAbort")
+
+	if r.cleanup(ctx, false, ibu) {
+		r.Log.Info("Finished handleAbort")
+		return doNotRequeue(), nil
+	}
+	return requeueWithError(fmt.Errorf("failed handelAbort"))
+}
+
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) handleAbortFailure(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	return doNotRequeue(), nil
+}
+
+func (r *ImageBasedUpgradeReconciler) handleFinalize(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	r.Log.Info("Starting handleFinalize")
+
+	if r.cleanup(ctx, true, ibu) {
+		r.Log.Info("Finished handleFinalize")
+		return doNotRequeue(), nil
+	}
+	return requeueWithError(fmt.Errorf("failed handleFinalize"))
+}
+
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) handleFinalizeFailure(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+
+	// TODO actual steps
+	// If succeeds, return doNotRequeue
+	return doNotRequeue(), nil
+}
+
+// cleanup cleans stateroots, precache, backup, ibu files and OADP operator
+// returns true if all cleanup tasks were successful
+func (r *ImageBasedUpgradeReconciler) cleanup(
+	ctx context.Context, allUnbootedStateroots bool, ibu *lcav1alpha1.ImageBasedUpgrade) bool {
+	// try to clean up as much as possible and avoid returning when one of the cleanup tasks fails
+	// successful means that all the cleanup tasks completed without any error
+	successful := true
+
+	if err := r.cleanupStateroots(allUnbootedStateroots, ibu); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup stateroots")
+	}
+	if err := r.Precache.Cleanup(ctx); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup precaching resources")
+	}
+	if _, err := r.BackupRestore.CleanupBackups(ctx); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to cleanup backups")
+	}
+	if err := r.BackupRestore.DeleteOadpOperator(ctx, backuprestore.OadpNs); err != nil {
+		successful = false
+		r.Log.Error(err, "failed to delete OADP operator")
+	}
```

Comment:
@Missxiaoguo can we call run these cleanups even if backup has not been called?

Reply:
Yes, without oadp/backup, nothing to delete

Reply:
Probably not for the CleanupBackups. The function queries the backup CR, if the OADP is not installed and the CRD is not available, it will fail. Need to add a check in the function see https://github.com/Missxiaoguo/lifecycle-agent/blob/main/internal/backuprestore/backup.go#L276-L280